### PR TITLE
README: Add Error codes into the readme

### PR DESCRIPTION
**Pull request by iamniting**
Signed-off-by: Nitin Goyal <nigoyal@redhat.com>

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -124,6 +135,18 @@ message DisableVolumeReplicationResponse {
 }
 ```
 
+#### Error Scheme
+
+| Condition | gRPC Code | Description | Recovery Behavior |
+|-------------------|-------------------|-------------------|-------------------|
+| Missing required field | INVALID_ARGUMENT | Indicates that a required field is missing from the request. | Caller MUST fix the request by adding the missing required field before retrying. |
+| Source does not exist | NOT_FOUND | Indicates that the specified source does not exist. | Caller MUST verify that the source is correct, the source is accessible, and has not been deleted before retrying with exponential back off. |
+| Volume in use | FAILED_PRECONDITION | Indicates that the volume corresponding to the specified `volume_id` could not be Disabled due to failed precondition (for example volume is not enabled). | Caller SHOULD ensure that volume is enabled. |
```
Comment (Madhu-1) [inline README.md:144]: if the mirroring is not enabled we should return success as spec need to be idempotent

File: README.md [hunk 2]
```diff
@@ -216,3 +263,15 @@ message ResyncVolumeResponse{
...
 }
 ```
+
+#### Error Scheme
+
+| Condition | gRPC Code | Description | Recovery Behavior |
+|-------------------|-------------------|-------------------|-------------------|
+| Missing required field | INVALID_ARGUMENT | Indicates that a required field is missing from the request. | Caller MUST fix the request by adding the missing required field before retrying. |
+| Source does not exist | NOT_FOUND | Indicates that the specified source does not exist. | Caller MUST verify that the source is correct, the source is accessible, and has not been deleted before retrying with exponential back off. |
+| Volume in use | FAILED_PRECONDITION | Indicates that the volume corresponding to the specified volume_id could not be resynced due to failed precondition (for example mirroring is not enabled or the resync site is not available). | Caller SHOULD ensure that mirroring is enabled. |
```
Comment (Madhu-1) [inline README.md:273]:
```suggestion
Volume in use | FAILED_PRECONDITION | Indicates that the volume corresponding to the specified volume_id could not be resynced due to failed precondition (for example mirroring is not enabled or the image is in the demoted state). | Caller SHOULD ensure that mirroring is enabled and the image is demoted.
```

File: README.md [hunk 3]
```diff
@@ -97,6 +97,17 @@ message EnableVolumeReplicationResponse {
...
+| Source does not exist | NOT_FOUND | Indicates that the specified source does not exist. | Caller MUST verify that the source is correct, the source is accessible, and has not been deleted before retrying with exponential back off. |
+Operation pending for volume | ABORTED | Indicates that there is already an operation pending for the specified `volume_id`. In general the Cluster Orchestrator (CO) is responsible for ensuring that there is no more than one call "in-flight" per `volume_id` at a given time. However, in some circumstances, the CO MAY lose state (for example when the CO crashes and restarts), and MAY issue multiple calls simultaneously for the same `volume_id`. The Plugin, SHOULD handle this as gracefully as possible, and MAY return this error code to reject secondary calls. | Caller SHOULD ensure that there are no other calls pending for the specified `volume_id`, and then retry with exponential back off. |
+| Call not implemented | UNIMPLEMENTED | The invoked RPC is not implemented by the Plugin or disabled in the Plugin's current mode of operation. | Caller MUST NOT retry. |
```
Comment (Madhu-1) [inline README.md:107]: Unimplemented won't be applicable for enable and disable operations. need to be removed.

**Review by Madhu-1 — COMMENTED** [review]
Small nit.

File: README.md (hunk 1 above)
Comment (iamniting) [reply README.md:144]: yeah right, what can be precondition here?

Comment (Madhu-1) [reply README.md:144]: no precondition. we can remove the pre-condition here

**Comment by iamniting** [comment]
@Madhu-1 Pls take a look again

File: README.md (hunk 1 above)
Comment (Madhu-1) [reply README.md:135]: missed to remove precondiftion here?

File: README.md [hunk 4]
```diff
@@ -97,6 +90,15 @@ message EnableVolumeReplicationResponse {
 }
 ```
 
+#### Error Scheme
+
+| Condition | gRPC Code | Description | Recovery Behavior |
+|-------------------|-------------------|-------------------|-------------------|
+| Missing required field | INVALID_ARGUMENT | Indicates that a required field is missing from the request. | Caller MUST fix the request by adding the missing required field before retrying. |
+| Source does not exist | NOT_FOUND | Indicates that the specified source does not exist. | Caller MUST verify that the source is correct, the source is accessible, and has not been deleted before retrying with exponential back off. |
```
Comment (Madhu-1) [inline README.md:98]:
```suggestion

Volume does not exist | NOT_FOUND | Indicates that a volume corresponding to the specified volume_id does not exist. | Caller MUST verify that the volume_id is correct and that the volume is accessible and has not been deleted before retrying with exponential back off.
````

This need to be replaced in other places also.

File: README.md [hunk 5]
```diff
@@ -97,6 +90,15 @@ message EnableVolumeReplicationResponse {
 }
 ```
 
+#### Error Scheme
+
+| Condition | gRPC Code | Description | Recovery Behavior |
+|-------------------|-------------------|-------------------|-------------------|
+| Missing required field | INVALID_ARGUMENT | Indicates that a required field is missing from the request. | Caller MUST fix the request by adding the missing required field before retrying. |
```
Comment (sp98) [inline README.md:97]: do we need to pass in error code number as well ?

File: README.md [hunk 6]
```diff
@@ -97,6 +90,15 @@ message EnableVolumeReplicationResponse {
...
+Operation pending for volume | ABORTED | Indicates that there is already an operation pending for the specified `volume_id`. In general the Cluster Orchestrator (CO) is responsible for ensuring that there is no more than one call "in-flight" per `volume_id` at a given time. However, in some circumstances, the CO MAY lose state (for example when the CO crashes and restarts), and MAY issue multiple calls simultaneously for the same `volume_id`. The Plugin, SHOULD handle this as gracefully as possible, and MAY return this error code to reject secondary calls. | Caller SHOULD ensure that there are no other calls pending for the specified `volume_id`, and then retry with exponential back off. |
+| Not authenticated | UNAUTHENTICATED | The invoked RPC does not carry secrets that are valid for authentication. | Caller SHALL either fix the secrets provided in the RPC, or otherwise regalvanize said secrets such that they will pass authentication by the Plugin for the attempted RPC, after which point the caller MAY retry the attempted RPC. |
+| Error in Unknown | UNKNOWN | Indicates that a unknown error is generated | Caller MUST study the logs before retrying |
```
Comment (sp98) [inline README.md:101]:
```suggestion
| Error is Unknown | UNKNOWN | Indicates that a unknown error is generated | Caller MUST study the logs before retrying |
```

File: README.md [hunk 7]
```diff
@@ -154,6 +165,17 @@ message PromoteVolumeResponse{
 }
 ```
 
+#### Error Scheme
+
+| Condition | gRPC Code | Description | Recovery Behavior |
+|-------------------|-------------------|-------------------|-------------------|
+| Missing required field | INVALID_ARGUMENT | Indicates that a required field is missing from the request. | Caller MUST fix the request by adding the missing required field before retrying. |
+| Volume does not exist | NOT_FOUND | Indicates that a volume corresponding to the specified `volume_id` does not exist. | Caller MUST verify that the `volume_id` is correct and that the volume is accessible and has not been deleted before retrying with exponential back off. |
+| Volume in use | FAILED_PRECONDITION | Indicates that the volume corresponding to the specified `volume_id` could not be promoted due to failed precondition (for example mirroring is not enabled). | Caller SHOULD ensure that mirroring is enabled. |
```
Comment (sp98) [inline README.md:174]: could this be `volume not available` condition rather than `volume in use`?

File: README.md (hunk 5 above)
Comment (iamniting) [reply README.md:97]: @Madhu-1 Can you pls confirm?

File: README.md (hunk 7 above)
Comment (iamniting) [reply README.md:174]: It will be ambiguous IMO, people can think that volume is not available in the cluster

File: README.md (hunk 5 above)
Comment (Madhu-1) [reply README.md:97]: good to document the error codes as well. documented here https://grpc.github.io/grpc/core/md_doc_statuscodes.html

File: README.md (hunk 7 above)
Comment (Madhu-1) [reply README.md:174]: volume not available/not found is documented in the second point? and also it cannot be volume in use rather it should be Volume is not replicated?

Comment (iamniting) [reply README.md]: done

File: README.md (hunk 5 above)
Comment (iamniting) [reply README.md]: done

**Review by Madhu-1 — APPROVED** [review]

**Review by sp98 — APPROVED** [review]

**Pull request merged**